src/main/java/dev/ecattez/shahmat/board/Piece.java

```diff
@@ -0,0 +1,62 @@
+package dev.ecattez.shahmat.board;
+
+import java.util.Objects;
+
+public abstract class Piece implements Typed, Colored, Oriented {
+
+    public final PieceType type;
+    public final PieceColor color;
+
+    public Piece(PieceType type, PieceColor color) {
+        this.type = type;
+        this.color = color;
+    }
+
+    @Override
+    public boolean isOfType(PieceType type) {
+        return this.type.equals(type);
+    }
+
+    @Override
+    public boolean isOfColor(PieceColor color) {
+        return this.color.equals(color);
+    }
+
+    @Override
+    public Orientation orientation() {
+        return color.accept(new PieceColorVisitor<Orientation>() {
+            @Override
+            public Orientation visitBlack() {
+                return Orientation.BLACK_VIEW;
+            }
+
+            @Override
+            public Orientation visitWhite() {
+                return Orientation.WHITE_VIEW;
+            }
+        });
+    }
+
+    public boolean isAlly(Piece piece) {
+        return piece != null && piece.isOfColor(this.color);
+    }
+
+    public boolean isOpponent(Piece piece) {
+        return !isAlly(piece);
```

Comment:
Si piece est `null`, isOpponent est vrai, ce qui est faux.

Plutôt faire : `return piece != null && piece.isOfColor(this.color.opposite())`